### PR TITLE
Adopt idiomatic error handling for executor.

### PR DESCRIPTION
**Pull request by zonr**
Stop using `panic`/`recover` to throw errors. Instead, stick to Go's
style for handling error. Function reports errors by returning an Go's
`error` value (instead of passing with `panic`). Errors are explicitly
checked or bypassed at where they might occur (instead of checking the
return value from `recover`).

Discussion: #28
Related: #44
Invalidate: #243, #253
Close: #402

### CODE REVIEW

**Comment by atombender** [comment]
What's preventing this from being merged?

**Comment by Fontinalis** [comment]
@egonelbre Could you take a look at this?

File: executor.go [hunk 1]
```diff
@@ -42,20 +42,19 @@ func Execute(p ExecuteParams) (result *Result) {
...
 			result.Errors = append(result.Errors, extErrs...)
 		}
-
 		addExtensionResults(&p, result)
 	}()
 
 	resultChannel := make(chan *Result, 2)
 
-	go func() {
+	go func(out chan<- *Result, done <-chan struct{}) {
```
Comment (egonelbre) [inline executor.go]: nit: no need to pass them in explicitly it's easier to see when using `resultChannel` directly.

File: executor.go [hunk 2]
```diff
@@ -42,20 +42,19 @@ func Execute(p ExecuteParams) (result *Result) {
...
 		result := &Result{}
 
 		defer func() {
```
Comment (egonelbre) [inline executor.go]: by not using recover `defer func() { resultChannel <- result }()`  will suffice

File: executor.go [hunk 3]
```diff
@@ -79,11 +77,12 @@ func Execute(p ExecuteParams) (result *Result) {
 			Root:             p.Root,
 			Operation:        exeContext.Operation,
 		})
-	}()
+		return
```
Comment (egonelbre) [inline executor.go]: unneeded extra return

**Comment by atombender** [comment]
Can this get some attention?

**Comment by atombender** [comment]
@Fontinalis @egonelbre What's left for this to be merged? It's been 7 months, and this is bugging us.

**Comment by egonelbre** [comment]
@Fontinalis I made the readability changes I commented in https://github.com/egonelbre/graphql/commit/3e383bec6e59522df4bed6219cf4f2ae11c0fe73, but I have no clue how to get these changes into this PR.

We could always merge as is and I can a do separate PR against.

**Comment by jpascal** [comment]
How long time we must wait to this feature?

**Pull request closed without merging**